src/System.Buffers.Primitives/System/Collections/Position.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Buffers;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Sequences
+{
+    public readonly struct Position : IEquatable<Position>
+    {
+        readonly object _segment;
+        readonly int _index;
+
+        public Position(object segment, int index)
+        {
+            _segment = segment;
+            _index = index;
+        }
+
+        public Position(object segment)
+        {
+            _segment = segment;
+            _index = 0;
+        }
+
+        public object Segment => _segment;
+        public int Index => _index;
+
+        public static explicit operator int(Position position) => position._index;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public (T segment, int index) Get<T>()
```

Comment:
Do we expect consumers to check if the segment is non-null/default?

Reply:
Depends on what they use the position for.

Reply:
I think in most cases they either use it or not. If they don't use it, direct access to Index is enough. If they need both then I prefer throwing to returning default.

Reply:
Sure, we can play with this. But I still want to do an experiment with changing the check to typeof(T) == typeof(Segment). I will try to do it after the new year. If if does not work, we can remove this method.